lib/ABI/Definition.cpp

```diff
@@ -0,0 +1,317 @@
+    // Skip ABIs that do not allow returning big values.
+    // They do not benefit from this check.
+    return true;
+  }
+
+  // Make sure the architecture of of the register is as expected.
+  const auto Architecture = model::ABI::getArchitecture(D.ABI());
+  const model::Register::Values RVLR = D.ReturnValueLocationRegister();
+  if (model::Register::getArchitecture(RVLR) != Architecture)
+    return false;
+
+  if (isVectorRegister(D.ReturnValueLocationRegister())) {
+    // Vector register used as the return value locations are not supported.
+    return false;
+  } else if (llvm::is_contained(D.CalleeSavedRegisters(),
+                                D.ReturnValueLocationRegister())) {
+    // Using callee saved register as a return value location doesn't make
+    // much sense: filter those out.
+    return false;
+  } else {
+    // The return value location register can optionally also be the first
+    // GPRs, but only the first one.
+    const auto &GPRs = D.GeneralPurposeArgumentRegisters();
+    const auto Iterator = llvm::find(GPRs, D.ReturnValueLocationRegister());
+    if (Iterator != GPRs.end() && Iterator != GPRs.begin())
+      return false;
+  }
+
+  return true;
+}
+
+using RegisterVector = std::vector<model::Register::Values>;
+
+template<typename RegisterType>
+bool compatibilityHelper(const SortedVector<RegisterType> &UsedRegisters,
+                         const RegisterVector &AllowedRegisters,
+                         model::Architecture::Values Architecture) {
+  for (const RegisterType &Register : UsedRegisters) {
+    // Verify the architecture of used registers.
+    if (model::Register::getArchitecture(Register.Location()) != Architecture)
+      revng_abort();
+  }
+
+  // Verify that every used register is also allowed.
+  for (const RegisterType &Register : UsedRegisters)
+    if (llvm::count(AllowedRegisters, Register.Location()) != 1)
+      return false;
+
+  return true;
+}
+
+namespace abi {
+
+bool Definition::isValid() const {
+  if (ABI() == model::ABI::Invalid)
+    return false;
+
+  const auto Architecture = model::ABI::getArchitecture(ABI());
+  if (!verifyRegisters(GeneralPurposeArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(GeneralPurposeReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(CalleeSavedRegisters(), Architecture))
+    return false;
+
+  if (!verifyReturnValueLocation(*this))
+    return false;
+
+  // TODO: extend to verify `Types` once it's more substantial,
+  //       i.e. holds vector argument information.
+
+  return true;
+}
+
+template<typename... Ts>
+using SV = SortedVector<Ts...>;
+
+template<typename AR, typename RVR>
+bool Definition::checkCompatibility(const SV<AR> &ArgumentRegisters,
```

Comment:
Actually, I'm thinking about reworking this into taking a `model::RawFunctionType` straight-up, but then I'm even more hesitant about naming. I have no idea how to clearly convey the fact that "returning true does not guarantees compatibility".

Maybe `ABI.isIncompatible(MyFunction)`? What do you think?

Reply:
> I feel like this function could be split in two and the second half could just be a template function that checks if a set-like object 1 is container within a set-like object 2.

Yes, it could, but that would also mean unnecessary sorts: we lose performance for nothing considering that both of these containers are always small. It's cheaper to just do an exhaustive search each time over doing two sorts into a "fully contained" check.

Reply:
Something like this, what do you think? https://github.com/revng/revng/pull/232/commits/c641315f64e8af81b3fb7ffc26ccb90b2f521a30.

Reply:
(ping)

Reply:
> we lose performance

Early optimization. This will never be the bottleneck: let's go for clarity over performance.

> I feel like this function could be split in two and the second half could just be a template function that checks if a set-like object 1 is container within a set-like object 2.

I still think this is the way to go.

Reply:
👌